File: ios/Runner/Plugins/HealthDataPlugin.swift
//
//  HealthDataPlugin.swift
//  Runner
//
//  Created by Danijel Tolj on 5. 2. 2024..
//

import Foundation
import Flutter

extension FlutterError: Error {}

class HealthDataPlugin: HealthDataHostApi {
    func getSteps(timestampFrom: Int64, timestampTo: Int64, completion: @escaping (Result<[StepsData], Error>) -> Void) {
        // competion(.success(data))
        // completion(.failure())
    }
    
    private let flutterApi: HealthDataFlutterApi
    private var timer: Timer?
    
    init(binaryMessenger: FlutterBinaryMessenger) {
        self.flutterApi = HealthDataFlutterApi(binaryMessenger: binaryMessenger)
        HealthDataHostApiSetup.setUp(binaryMessenger: binaryMessenger, api: self)
        self.timer = Timer.scheduledTimer(timeInterval: 1.0, target: self, selector: #selector(timerTask), userInfo: nil, repeats: true)
    }
    
    @objc func timerTask() {
        let timeSeriesData = TimeSeriesData(
            timestamp: Int64(Date().timeIntervalSince1970 * 1000),
            data: Int64.random(in: 60...70)
        )
        self.flutterApi.onHeartRateAdded(data: timeSeriesData) { _ in }
    }
    
    func dispose() {
        timer?.invalidate()
        timer = nil
    }
    
    func getHeartRate(from: Int64, to: Int64, completion: @escaping (Result<[TimeSeriesData], Error>) -> Void) {
        // you can use the parameters passed from flutter method call here
        //let fromDate = Date(timeIntervalSince1970: TimeInterval(from) / 1000)
        //let toDate = Date(timeIntervalSince1970: TimeInterval(to) / 1000)
        
        
        // Fetch some health data
        let data = [
            TimeSeriesData(timestamp: Int64(Date().timeIntervalSince1970 * 1000) - 1000, data: Int64.random(in: 60...70)),
            TimeSeriesData(timestamp: Int64(Date().timeIntervalSince1970 * 1000) - 2000, data: Int64.random(in: 60...70)),
            TimeSeriesData(timestamp: Int64(Date().timeIntervalSince1970 * 1000) - 3000, data: Int64.random(in: 60...70))
        ]
        
        completion(.success(data))
    }
}

File: android/app/src/main/kotlin/com/example/pigeon_poc/plugin_handlers/HealthDataPlugin.kt
package com.example.pigeon_poc.plugin_handlers

import HealthDataFlutterApi
import HealthDataHostApi
import StepsData
import TimeSeriesData
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import java.time.Instant
import java.util.Timer
import java.util.TimerTask
import kotlin.random.Random

class HealthDataPlugin(binaryMessenger: BinaryMessenger) : HealthDataHostApi {
    private val _flutterApi: HealthDataFlutterApi = HealthDataFlutterApi(binaryMessenger)

    private val timer = Timer()
    private val handler = android.os.Handler(Looper.getMainLooper())

    private val task = object : TimerTask() {
        override fun run() {
            // Use Handler to execute UI-related task on the main thread
            handler.post {
                _flutterApi.onHeartRateAdded(
                    TimeSeriesData(
                        Instant.now().toEpochMilli(), Random.nextInt(60, 70 + 1).toLong()
                    )
                ) {}
            }
        }
    }

    init {
        HealthDataHostApi.setUp(binaryMessenger, this)
        timer.scheduleAtFixedRate(task, 0, 1000)
    }

    fun dispose() {
        timer.cancel()
    }

    override fun getHeartRate(
        from: Long,
        to: Long,
        callback: (Result<List<TimeSeriesData>>) -> Unit
    ) {
        // you can use the parameters passed from flutter method call here
        val from = Instant.ofEpochMilli(from)
        val to = Instant.ofEpochMilli(to)


        // Fetch some health data
        val data = listOf(
            TimeSeriesData(
                timestamp = Instant.now().toEpochMilli() + 1000,
                data = Random.nextInt(60, 70).toLong()
            ),
            TimeSeriesData(
                timestamp = Instant.now().toEpochMilli() + 2000,
                data = Random.nextInt(60, 70).toLong()
            ),
            TimeSeriesData(
                timestamp = Instant.now().toEpochMilli() + 3000,
                data = Random.nextInt(60, 70).toLong()
            )
        )
        callback(Result.success(data))
    }



    override fun getSteps(
        timestampFrom: Long,
        timestampTo: Long,
        callback: (Result<List<StepsData>>) -> Unit
    ) {
        TODO("Not yet implemented")
    }

}